### PR TITLE
tree-wide: replace string functions with fundamental functions

### PR DESCRIPTION
**Pull request by SiliconA-Z**

### CODE REVIEW

File: src/libsystemd/sd-bus/sd-bus.c [hunk 1]
```diff
@@ -649,7 +649,7 @@ static int parse_address_key(const char **p, const char *key, char **value) {
 
         if (key) {
                 l = strlen(key);
-                if (strncmp(*p, key, l) != 0)
+                if (!strneq(*p, key, l))
```
Comment (yuwata) [inline src/libsystemd/sd-bus/sd-bus.c]: When you touch this, please use `startswith()`.

File: src/libsystemd/sd-bus/sd-bus.c [hunk 2]
```diff
@@ -4121,7 +4121,7 @@ _public_ int sd_bus_path_decode_many(const char *path, const char *path_template
                 /* verify everything until the next '%' matches verbatim */
                 sep = strchrnul(template_pos, '%');
                 length = sep - template_pos;
-                if (strncmp(path_pos, template_pos, length))
+                if (!strneq(path_pos, template_pos, length))
```
Comment (yuwata) [inline src/libsystemd/sd-bus/sd-bus.c:4127]: Also here.

File: src/libsystemd/sd-bus/sd-bus.c [hunk 3]
```diff
@@ -4142,8 +4142,7 @@ _public_ int sd_bus_path_decode_many(const char *path, const char *path_template
 
                 /* verify the suffixes match */
                 sep = strchrnul(path_pos, '/');
-                if (sep - path_pos < (ssize_t)length ||
-                    strncmp(sep - length, template_pos, length))
+                if (sep - path_pos < (ssize_t) length || !strneq(sep - length, template_pos, length))
```
Comment (yuwata) [inline src/libsystemd/sd-bus/sd-bus.c]: Also here.

File: src/shared/efi-api.c [hunk 4]
```diff
@@ -453,13 +453,13 @@ int efi_get_boot_options(uint16_t **ret_options) {
         FOREACH_DIRENT(de, dir, return -errno) {
                 int id;
 
-                if (strncmp(de->d_name, "Boot", 4) != 0)
+                if (!strneq(de->d_name, "Boot", 4))
```
Comment (yuwata) [inline src/shared/efi-api.c]: Also here.

File: src/shared/fstab-util.c [hunk 5]
```diff
@@ -232,13 +232,14 @@ int fstab_filter_options(
                         }
 
                         NULSTR_FOREACH(name, names) {
-                                if (end < word + strlen(name))
+                                size_t length = strlen(name);
+                                if (end < word + length)
                                         continue;
-                                if (!strneq(word, name, strlen(name)))
+                                if (!strneq(word, name, length))
```
Comment (yuwata) [inline src/shared/fstab-util.c]: Also here.

File: src/udev/udev-builtin-path_id.c [hunk 6]
```diff
@@ -186,7 +186,7 @@ static sd_device *handle_scsi_sas(sd_device *parent, char **path) {
                 return NULL;
 
         /* Check if we are simple disk */
-        if (strncmp(phy_count, "1", 2) != 0)
+        if (!streq(phy_count, "1"))
```
Comment (yuwata) [inline src/udev/udev-builtin-path_id.c:189]: This is OK.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/libsystemd/sd-bus/sd-bus.c (hunk 3 above)
Comment (AtariDreams) [reply src/libsystemd/sd-bus/sd-bus.c:4145]: Ditto what I said earlier.

File: src/libsystemd/sd-bus/sd-bus.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd/sd-bus/sd-bus.c:4127]: Indeed. Please ignore the comment here.

File: src/shared/fstab-util.c [hunk 7]
```diff
@@ -232,14 +232,14 @@ int fstab_filter_options(
                         }
 
                         NULSTR_FOREACH(name, names) {
-                                if (end < word + strlen(name))
-                                        continue;
-                                if (!strneq(word, name, strlen(name)))
+                                char *match;
```
Comment (YHNdnzj) [inline src/shared/fstab-util.c]: const

File: src/shared/efi-api.c [hunk 8]
```diff
@@ -453,13 +453,14 @@ int efi_get_boot_options(uint16_t **ret_options) {
...
 
-                if (strncmp(de->d_name, "Boot", 4) != 0)
+                if (!startswith(de->d_name, "Boot"))
                         continue;
 
                 if (strlen(de->d_name) != 45)
                         continue;
 
-                if (strcmp(de->d_name + 8, EFI_GLOBAL_VARIABLE_STR("")) != 0)  /* generate variable suffix using macro */
+                if (!startswith(de->d_name + 8,
```
Comment (YHNdnzj) [inline src/shared/efi-api.c]: Why startswith?

File: src/libsystemd/sd-bus/sd-bus.c (hunk 3 above)
Comment (AZero13) [reply src/libsystemd/sd-bus/sd-bus.c]: @yuwata That cannot apply here due to the processing of the string being into pieces, as in we are not looking for only a null terminator, but dividing a string by the slashes as well.

**Comment by AtariDreams** [comment]
@yuwata Fixed!

File: src/basic/env-util.c [hunk 9]
```diff
@@ -244,9 +244,9 @@ static bool env_match(const char *t, const char *pattern) {
                 return true;
 
         if (!strchr(pattern, '=')) {
-                size_t l = strlen(pattern);
+                t = startswith(t, pattern);
 
-                return strneq(t, pattern, l) && t[l] == '=';
+                return t && *t == '=';
```
Comment (yuwata) [inline src/basic/env-util.c:249]: Not a matter, but a shorter version:
`return strneq_ptr(startswith(t, pattern), "=", 1);`

File: src/basic/string-util.c [hunk 10]
```diff
@@ -1509,24 +1509,24 @@ ssize_t strlevenshtein(const char *x, const char *y) {
...
+        char *result = NULL;
+        for (;;) {
+                char *p = strstr(haystack, needle);
+                if (p == NULL) {
+                        break;
+                }
+                result = p;
+                haystack = result + 1;
+        }
+        return result;
```
Comment (yuwata) [inline src/basic/string-util.c]: How about the following?
```c
for (const char *p = strstr(haystack, needle), *q; p; p = q) {
        q = strstr(p + 1, needle);
        if (!q)
                return (char*) p;
}
return NULL;
```

File: src/libsystemd/sd-bus/sd-bus.c [hunk 11]
```diff
@@ -4121,35 +4118,34 @@ _public_ int sd_bus_path_decode_many(const char *path, const char *path_template
...
+                if (!*sep)
                         break;
 
                 /* We found the next '%' character. Everything up until here
                  * matched. We now skip ahead to the end of this label and make
                  * sure it matches the tail of the label in the path. Then we
                  * decode the string in-between and save it for later use. */
 
-                ++template_pos; /* skip over '%' */
+                template_pos = sep + 1; /* skip over '%' */
```
Comment (yuwata) [inline src/libsystemd/sd-bus/sd-bus.c]: Please drop the change. `path_pos` and `template_pos` are both kind of cursors on `path` and `template`, and these should be handled in the same way.

File: src/shared/fstab-util.c [hunk 12]
```diff
@@ -187,7 +187,7 @@ int fstab_filter_options(
                                 if (!x)
                                         continue;
                                 /* Match name, but when ret_values, only when followed by assignment. */
-                                if (*x == '=' || (!ret_values && *x == '\0')) {
+                                if (*x == '=' || (*x == '\0' && !ret_values)) {
```
Comment (yuwata) [inline src/shared/fstab-util.c]: Please drop meaningless change.

File: src/shared/fstab-util.c [hunk 13]
```diff
@@ -232,23 +232,21 @@ int fstab_filter_options(
...
-                                if (end < word + strlen(name))
-                                        continue;
-                                if (!strneq(word, name, strlen(name)))
+                                x = startswith(word, name);
+                                if (!x || x > end)
                                         continue;
 
-                                /* We know that the string is NUL terminated, so *x is valid */
-                                x = word + strlen(name);
+                                /* We know that the string is NUL terminated, so *match is valid */
```
Comment (yuwata) [inline src/shared/fstab-util.c]: `*match` -> `*x`

File: src/shared/fstab-util.c [hunk 14]
```diff
@@ -232,23 +232,21 @@ int fstab_filter_options(
...
                                 if (IN_SET(*x, '\0', '=', ',')) {
                                         namefound = name;
                                         break;
                                 }
                         }
 
-                        if (*end)
-                                word = end + 1;
-                        else
+                        if (!*end)
```
Comment (yuwata) [inline src/shared/fstab-util.c]: Our basic rule is that implicit cast to boolean is only allowed for pointers. So, I'd like to make this
`if (*end == '\0')`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Please split out the change for `strrstr()` to another commit.

File: src/basic/string-util.c [hunk 15]
```diff
@@ -1509,24 +1509,20 @@ ssize_t strlevenshtein(const char *x, const char *y) {
...
                 return NULL;
 
-        l = strlen(needle);
-
         /* Special case: for the empty string we return the very last possible occurrence, i.e. *after* the
          * last char, not before. */
-        if (l == 0)
-                return strchr(haystack, 0);
+        if (*needle == 0)
+                return (char *) haystack + strlen(haystack);
```
Comment (poettering) [inline src/basic/string-util.c]: why change this?

Comment (AZero13) [reply src/basic/string-util.c]: To avoid having to call strlen

Comment (yuwata) [reply src/basic/string-util.c]: Please drop the change for the `return` line. You do not answer the question, and in the changed line you call `strlen()`...

File: src/basic/process-util.c [hunk 16]
```diff
@@ -1074,14 +1071,18 @@ int getenv_for_pid(pid_t pid, const char *field, char **ret) {
...
+                        value = strdup(match + 1);
                         if (!value)
                                 return -ENOMEM;
 
                         *ret = value;
                         return 1;
                 }
+
+                if (sum > ENVIRONMENT_BLOCK_MAX) /* Give up searching eventually */
+                        return -ENOBUFS;
```
Comment (yuwata) [inline src/basic/process-util.c]: Why moved? I'd like to keep this at the original place.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]

**Pull request merged**